JSWebAssemblyInstance should be allocated out of a new PreciseSubspace.

kmiller68 opened the pull request:
#### b8c3632cf76735ab4151bd90fcd370d18d948c5f
<pre>
JSWebAssemblyInstance should be allocated out of a new PreciseSubspace.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287518">https://bugs.webkit.org/show_bug.cgi?id=287518</a>
<a href="https://rdar.apple.com/144647030">rdar://144647030</a>

Reviewed by Yusuke Suzuki.

Previously we added a new mode, preciseOnly, to IsoSubspace. However, this was unnecessarily complicated
and wasteful. Additionally, it was confusing since most places assumed IsoSubspaces always allocated
objects of the same size.

This change adds a new PreciseSubspace that always allocates objects into a PreciseAllocation and factors
out most of this logic. In a subsequent patch I&apos;ll remove the old, now dead, preciseOnly logic.

I tried to have the heap account for IsoSubspace&apos;s lower tier precise as part of it&apos;s capacity
but that seemed to cause significant performance regressions across the board both in memory and runtime.
So I&apos;m going to skip that for now with a FIXME to correct it later.

* JSTests/wasm/stress/stress-JSWebAssemblyInstance-allocations.js: Added.
(catch):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/heap/CompleteSubspace.cpp:
(JSC::CompleteSubspace::tryAllocateSlow):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::deleteAllCodeBlocks):
(JSC::Heap::updateAllocationLimits):
(JSC::Heap::webAssemblyInstanceSpaceSlow):
* Source/JavaScriptCore/heap/Heap.h:
(JSC::Heap::webAssemblyInstanceSpace):
* Source/JavaScriptCore/heap/IsoSubspace.cpp:
(JSC::IsoSubspace::tryAllocatePreciseOrLowerTierPrecise):
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::registerPreciseAllocation):
* Source/JavaScriptCore/heap/MarkedSpace.h:
* Source/JavaScriptCore/heap/PreciseAllocation.h:
* Source/JavaScriptCore/heap/PreciseSubspace.cpp: Added.
(JSC::PreciseSubspace::PreciseSubspace):
(JSC::PreciseSubspace::didResizeBits):
(JSC::PreciseSubspace::didRemoveBlock):
(JSC::PreciseSubspace::didBeginSweepingToFreeList):
(JSC::PreciseSubspace::tryAllocate):
(JSC::PreciseSubspace::allocate):
* Source/JavaScriptCore/heap/PreciseSubspace.h: Added.
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::updateStackLimits):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::webAssemblyInstanceSpace):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:

Canonical link: <a href="https://commits.webkit.org/290699@main">https://commits.webkit.org/290699@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0bd3cfc4c2685299970e4863be1b6c2d2f80671

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90658 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10195 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45604 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95673 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41444 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10589 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18512 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69779 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27323 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93659 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8091 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82222 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50121 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7846 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36604 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40573 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83471 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78161 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37666 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97503 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89438 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17853 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13123 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78810 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18112 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78046 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78007 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/19268 "The change is no longer eligible for processing.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22432 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/21060 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14318 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17862 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23205 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111927 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17601 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21057 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19385 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/heap/Heap.cpp:
```diff
@@ -1105,7 +1105,8 @@ void Heap::deleteAllCodeBlocks(DeleteAllCodeEffort effort)
         // it uses a callee check, but then it will call into dead code.
-        HeapIterationScope heapIterationScope(*this);
+
+        // PreciseAllocations are always eagerly swept so we don't have to worry about handling instances pending destruction thus need a HeapIterationScope
```
That's awesome

Constellation reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/heap/IsoSubspace.cpp:
```diff
@@ -87,32 +87,28 @@ void IsoSubspace::didBeginSweepingToFreeList(MarkedBlock::Handle* block)
 
 void* IsoSubspace::tryAllocatePreciseOrLowerTierPrecise(size_t size)
```
We can also remove preciseOnly mode code from IsoSubspace.

kmiller68 replied:
Yeah, I was going to do that in a follow up.

kmiller68 committed:
JSWebAssemblyInstance should be allocated out of a new PreciseSubspace.
https://bugs.webkit.org/show_bug.cgi?id=287518
rdar://144647030

Reviewed by Yusuke Suzuki.

Previously we added a new mode, preciseOnly, to IsoSubspace. However, this was unnecessarily complicated
and wasteful. Additionally, it was confusing since most places assumed IsoSubspaces always allocated
objects of the same size.

This change adds a new PreciseSubspace that always allocates objects into a PreciseAllocation and factors
out most of this logic. In a subsequent patch I'll remove the old, now dead, preciseOnly logic.

I tried to have the heap account for IsoSubspace's lower tier precise as part of it's capacity
but that seemed to cause significant performance regressions across the board both in memory and runtime.
So I'm going to skip that for now with a FIXME to correct it later.

* JSTests/wasm/stress/stress-JSWebAssemblyInstance-allocations.js: Added.
(catch):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/heap/CompleteSubspace.cpp:
(JSC::CompleteSubspace::tryAllocateSlow):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::deleteAllCodeBlocks):
(JSC::Heap::updateAllocationLimits):
(JSC::Heap::webAssemblyInstanceSpaceSlow):
* Source/JavaScriptCore/heap/Heap.h:
(JSC::Heap::webAssemblyInstanceSpace):
* Source/JavaScriptCore/heap/IsoSubspace.cpp:
(JSC::IsoSubspace::tryAllocatePreciseOrLowerTierPrecise):
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::registerPreciseAllocation):
* Source/JavaScriptCore/heap/MarkedSpace.h:
* Source/JavaScriptCore/heap/PreciseAllocation.h:
* Source/JavaScriptCore/heap/PreciseSubspace.cpp: Added.
(JSC::PreciseSubspace::PreciseSubspace):
(JSC::PreciseSubspace::didResizeBits):
(JSC::PreciseSubspace::didRemoveBlock):
(JSC::PreciseSubspace::didBeginSweepingToFreeList):
(JSC::PreciseSubspace::tryAllocate):
(JSC::PreciseSubspace::allocate):
* Source/JavaScriptCore/heap/PreciseSubspace.h: Added.
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::updateStackLimits):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::webAssemblyInstanceSpace):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:

Canonical link: https://commits.webkit.org/290699@main
